### PR TITLE
feat(cat-voices): keychain and session

### PR DESCRIPTION
**Pull request by dtscalac**
# Description

Adds support for keychain which stores the seed phrase and updates session bloc to lock/unlock the session.

Next PRs will:
- add support for lock/unlock
- add support for session expiration (currently once unlocked it stays unlocked)

Clicking on these links will put the application in different states with mocked data:

Screenshot 2024-10-14 at 08 40 08

```dart
  /// Temporary implementation for testing purposes.
  User get _dummyUser => const User(name: 'Account');

  /// Temporary implementation for testing purposes.
  SeedPhrase get _dummySeedPhrase => SeedPhrase.fromMnemonic(
        'few loyal swift champion rug peace dinosaur'
        ' erase bacon tone install universe',
      );

  /// Temporary implementation for testing purposes.
  LockFactor get _dummyUnlockFactor => const PasswordLockFactor('Test1234');
```

## Related Issue(s)

Refers #813

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

**Review by damian-molinski — CHANGES_REQUESTED** [review]

File: catalyst_voices/packages/catalyst_voices_services/lib/src/keychain/keychain_service.dart [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+const _seedPhraseKey = 'keychain_seed_phrase';
+
+/// The keychain which holds the user's [SeedPhrase] and allows
+/// to derive the [Ed25519KeyPair] related to the [AccountRole].
+///
+// TODO(dtscalac): in the future when key derivation algorithm spec
+// will become stable consider to store derived keys instead of deriving
+// them each time they are needed.
+class KeychainService {
```
[the remark this answers is not in the document]
Comment (dtscalac) [reply catalyst_voices/packages/catalyst_voices_services/lib/src/keychain/keychain_service.dart:16]: Renamed to Keychain

File: catalyst_voices/lib/app/view/app.dart [hunk 2]
```diff
@@ -31,7 +31,8 @@ final class App extends StatelessWidget {
         create: (_) => Dependencies.instance.get<LoginBloc>(),
       ),
       BlocProvider<SessionBloc>(
-        create: (_) => Dependencies.instance.get<SessionBloc>(),
+        create: (_) => Dependencies.instance.get<SessionBloc>()
+          ..add(const RestoreSessionEvent()),
```
[the remark this answers is not in the document]
Comment (dtscalac) [reply catalyst_voices/lib/app/view/app.dart:35]: Converted it to stateful widget, moved the logic into initState()

File: catalyst_voices/packages/catalyst_voices_services/lib/src/storage/storage.dart [hunk 3]
```diff
@@ -11,35 +11,73 @@ import 'package:flutter/foundation.dart';
 /// Implementation may use local memory / filesystem or shared preferences or
 /// any other.
 abstract interface class Storage {
+  /// Returns true if a [String] with given [key] exists
+  /// in the storage and has non-null value.
+  ///
+  /// The method must be always callable,
+  /// despite the session being locked, keychain encrypted, etc.
+  Future<bool> containsString({required String key});
```
[the remark this answers is not in the document]
Comment (dtscalac) [reply catalyst_voices/packages/catalyst_voices_services/lib/src/storage/storage.dart:19]: fixed

File: catalyst_voices/packages/catalyst_voices_services/lib/src/keychain/keychain_service.dart [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    final hexEntropy = await _vault.readString(key: _seedPhraseKey);
+    if (hexEntropy != null) {
+      return SeedPhrase.fromHexEntropy(hexEntropy);
+    } else {
+      return null;
+    }
+  }
+
+  Future<void> _writeSeedPhrase(SeedPhrase seedPhrase) async {
+    await _vault.writeString(seedPhrase.hexEntropy, key: _seedPhraseKey);
```
[the remark this answers is not in the document]
Comment (dtscalac) [reply catalyst_voices/packages/catalyst_voices_services/lib/src/keychain/keychain_service.dart]: Ok, added todo

File: catalyst_voices/packages/catalyst_voices_services/lib/src/keychain/keychain_service.dart [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  Future<bool> get isUnlocked => _vault.isUnlocked;
+
+  /// Returns true if the keychain contains the seed phrase, false otherwise.
+  Future<bool> get hasSeedPhrase async => _hasSeedPhrase;
+
+  /// Initializes the keychain with [seedPhrase] and the [unlockFactor].
+  ///
+  /// In most cases the [unlockFactor] is going to be
+  /// an instance of a [PasswordLockFactor].
+  Future<void> init({
```
[the remark this answers is not in the document]
Comment (dtscalac) [reply catalyst_voices/packages/catalyst_voices_services/lib/src/keychain/keychain_service.dart]: Applied this suggestion

File: catalyst_voices/lib/app/view/app.dart [hunk 6]
```diff
@@ -31,7 +44,7 @@ final class App extends StatelessWidget {
         create: (_) => Dependencies.instance.get<LoginBloc>(),
       ),
       BlocProvider<SessionBloc>(
-        create: (_) => Dependencies.instance.get<SessionBloc>(),
+        create: (_) => _sessionBloc,
       ),
```
Comment (damian-molinski) [inline catalyst_voices/lib/app/view/app.dart:48]:
```dart
BlocProvider<SessionBloc>.value(_sessionBloc)
```

File: catalyst_voices/lib/app/view/app.dart [hunk 7]
```diff
@@ -4,20 +4,33 @@ import 'package:catalyst_voices_blocs/catalyst_voices_blocs.dart';
...
+  State<App> createState() => _AppState();
+}
+
+class _AppState extends State<App> {
+  final SessionBloc _sessionBloc = Dependencies.instance.get();
+
+  @override
+  void initState() {
+    super.initState();
+    _sessionBloc.add(const RestoreSessionEvent());
```
[the remark this answers is not in the document]
Comment (dtscalac) [reply catalyst_voices/lib/app/view/app.dart:26]: Added comment to the bloc field

**Review by damian-molinski — APPROVED** [review]
LGTM!

**Pull request merged**